### ISSUE TITLE
Injection when there is cyclic references fails

### ISSUE THREAD

**Issue by WaDelma** [FasterXML/jackson-databind#538]
When trying to debug problem in my project I was able reduce the problem to this simple case...
It seems that Jackson cannot handle injecting values via constructor when there is cyclic references.
The error disappears when I remove the cyclic reference or inject from either constructor.
```java
com.fasterxml.jackson.databind.deser.UnresolvedForwardReference: Could not resolve Object Id [1] (for [simple type, class test.Test$A]).
 at [Source: {"@id":1,"b":{"@id":2,"a":1}}; line: 1, column: 28]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectId(BeanDeserializerBase.java:1042)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromNumber(BeanDeserializerBase.java:1076)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:137)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:126)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:538)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:373)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1056)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:264)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithObjectId(BeanDeserializerBase.java:1028)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:122)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:538)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:373)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1056)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:264)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithObjectId(BeanDeserializerBase.java:1028)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:122)
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1269)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:896)
	at test.Test.main(Test.java:23)
```

```java
package test;

import java.io.IOException;

import com.fasterxml.jackson.annotation.JacksonInject;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import com.fasterxml.jackson.databind.InjectableValues;
import com.fasterxml.jackson.databind.ObjectMapper;

public class Test {
	public static void main(String[] args) {
		A a = new A(null);
		B b = new B(null);
		a.b = b;
		b.a = a;
		try {
			ObjectMapper mapper = new ObjectMapper();
			String json = mapper.writeValueAsString(a);
			InjectableValues.Std inject = new InjectableValues.Std();
			inject.addValue("i1", "e1");
			inject.addValue("i2", "e2");
			mapper.reader(inject).withType(A.class).readValue(json);
			System.out.println(json);
		} catch (IOException expection) {
			expection.printStackTrace();
		}
	}

	@JsonIdentityInfo(generator = ObjectIdGenerators.IntSequenceGenerator.class)
	public static class A {
		public B b;
		public A(@JacksonInject("i1") String injected) {}
	}

	@JsonIdentityInfo(generator = ObjectIdGenerators.IntSequenceGenerator.class)
	public static class B {
		public A a;
		public B(@JacksonInject("i2") String injected) {}
	}
}

```

**Comment by cowtowncoder** [comment]
Seems to be resolved for 2.8.0, possibly related to fixes to forwards-refs handling.

**Comment by cowtowncoder** [comment]
Note: Failing for Jackson 3.0.0-SNAPSHOT (`master`).
